### DISCUSSION TITLE
Help with PrismaMock which use db instead of mocking

### DISCUSSION THREAD

**Discussion by JeeremB** [prisma/prisma#12553] [Q&A] [answered]
Hello, I'm trying to use PrismaMock to do some unit tests on my project.
But, I'm facing an issue, I don't want to use a test db to do my tests so I've decided to use Prisma mock.
I have read the documentation about it (https://www.prisma.io/docs/guides/testing/unit-testing), I have put in place my tests following it. However, when I run my test, it applies it on the database.
Here is my code

the create role function in my service auth
```
@GrpcMethod("AuthService", "CreateRole")
async createRole(request: CreateRoleRequest): Promise<Role> {
  const roleToCreate: Role = {
    name: request.name,
    scopes: request.scopes,
    description: request.description,
  }

  return await prisma.role.create({
    data: roleToCreate
  })
}
```

the test
```
describe('createRoleMock', () => {
  it('should create a role', async () => {
    const role = {
      name: "default",
      description: "default role",
      scopes: ["user.read", "program.read"]
    };

    const appController = app.get<AppController>(AppController);

    prismaMock.role.create.mockResolvedValue(role)

    await expect(appController.createRole(role)).resolves.toEqual({
      name: "default",
      description: "default role",
      scopes: ["user.read", "program.read"]
    })
  });
});
```

the output:
![image](https://user-images.githubusercontent.com/97462049/160418941-a8caba59-001c-4484-81e9-0fc7ae222fee.png)
unique constraint failed on name because I use the name "default" to test my function and I already have a role called default in my db

I use the singleton solution of the documentation so I have the client.ts and the singleton.ts

Hope you can help me 🙏 (and sorry for my english 🇫🇷)
Jeerem 🚀

**Comment by austincrim** [comment, marked as answer]
Hey @JeeremB 👋 ,

Is it possible for you to share your repo with me? There are a number of moving pieces with the singleton mocking approach that it might just be easier this way 😅.

If you can't share the repo, here are some questions I have:
- Are you using Jest? 
- If so, did you add `singleton.ts` in `setupFilesAfterEnv` in the `jest.config.js` file?
- If not, what setup are you using?

**Reply by JeeremB** [reply]
Hey Austin,

Thank you for ur answer I found the solution to my issue. I've switched to the dependency injection mode and it works 👌

Have a nice day !

**Reply by G-Salviati** [reply]
Hello, I'm using the singleton pattern in nodejs and I have the same problem. The data are still retrieved from the database. Any suggestion?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
